budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,288 @@
+from datapackage_pipelines.wrapper import spew, ingest
+from datetime import datetime
+import itertools
+import logging
+
+value_fields = [
+    'net_expense_diff',
+    'gross_expense_diff',
+    'allocated_income_diff',
+    'commitment_limit_diff',
+    'personnel_max_diff'
+]
+
+
+def update_datapackage(datapackage):
+    key_fields = ['leading_item', 'req_code']
+    resource = datapackage['resources'][0]
+    fields = resource['schema']['fields']
+    new_fields = []
+    for field in fields:
+        name = field['name']
+        if name in key_fields:
+            new_fields.append(field)
+    new_fields.append({
+        'name': 'transaction_id',
+        'type': 'string'
+    })
+    resource['schema']['fields'] = new_fields
+    return datapackage
+
+
+def transfer_code(change):
+    return '%d/%02d-%03d' % (
+        change['year'],
+        change['leading_item'],
+        change['req_code']
+    )
+
+
+def powerset(iterable):
+    s = list(iterable)
+    return itertools.chain.from_iterable(
+        combinations(s, r)
+        for r in range(2, min(len(s) + 1, 6))
+    )
+
+
+def subsets(s):
+    return map(set, powerset(s))
+
+
+def combinations(iterable, r):
+    # combinations('ABCD', 2) --> AB AC AD BC BD CD
+    # combinations(range(4), 3) --> 012 013 023 123
+    def get_next(next_of, i):
+        ret = nexts.get(next_of, next_of + 1)
+        if ret > i + n - r:
+            ret = i + n - r
+        return ret
+
+    pool = tuple(iterable)
+    n = len(pool)
+    if r > n:
+        return
+    nexts = dict(zip(range(n - 1), range(1, n)))
+    indices = list(range(r))
+    while True:
+        found = (yield tuple(pool[i] for i in indices))
+        if found:
+            for i in indices:
+                target = nexts.get(i)
+                sources = (k for k, v in nexts.items() if v == i)
+                if target is not None:
+                    for src in sources:
+                        nexts[src] = target
+                else:
+                    for src in list(sources):
+                        del nexts[src]
+            indices[0] = get_next(indices[0], 0)
+            for j in range(1, r):
+                indices[j] = get_next(indices[j - 1], j)
+        for i in reversed(range(r)):
+            if indices[i] != i + n - r:
+                break
+        else:
+            return
+        indices[i] = get_next(indices[i], i)
+        for j in range(i + 1, r):
+            indices[j] = get_next(indices[j - 1], j)
+
+
+def get_changes(rows):
+    result = []
+
+    for row in rows:
+        change = dict(row)
+
+        # if change['year'] is None:
+        #     change['year'] = datetime.strptime(
+        #         change['date/approval'], '%d/%m/%Y').year
+
+        change['trcode'] = transfer_code(change)
+        change['_value'] = sum(change.get(x, 0) for x in value_fields)
+        for k, v in change.items():
+            if k.startswith('date'):
```

Comment:
yes, that's true.
So let me revise my previous comment.
- We should only attempt grouping for changes which have an approved date.
- In case there's no date, we should treat it as a group of one element (i.e. transaction id == request id).